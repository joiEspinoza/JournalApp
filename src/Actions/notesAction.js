import { db } from "../FireBase/fireBaseConfig";
import { HloadNotes } from "../Helpers/HloadNotes";
import types from "../Types/types";
import Swal from "sweetalert2";
import { HFileUpload } from "../Helpers/HFileUpload";

/////

const starNewNote = () =>
{
    return async ( dispatch, getState ) => {

        try 
        {
            const { uid } = getState().auth;

            const newNote = 
            {
                title : "",
                body : "",
                date : new Date().getTime()
            };
    
            const docRef = await db.collection( `${ uid }/journal/notes` ).add( newNote );
    
            dispatch( activeNote( docRef.id, newNote ) );
            
            dispatch( addNewNote( docRef.id, newNote ) );
        } 
        catch (error) 
        {
            console.log(error);
        }
      

    };
};

/////

const activeNote = ( id, note ) =>( { type : types.notesActive, payload : { id, ...note } } );

///////

const addNewNote = ( id, note ) => ( { type : types.notesAddNew, payload : { id, ...note } } );

//////

const startLoadingNotes = ( uid ) =>
{
    return async ( dispatch ) => 
    {
        try 
        {
            const notes = await HloadNotes( uid );

            dispatch( setNotes( notes ) );
        } 
        catch (error) 
        {
            console.log( error );
        }
     
    };
};

///////

const setNotes = ( notes ) => ( { type : types.notesLoad, payload : notes } );

//////

const saveNote = ( note ) =>
{
    return async ( dispatch, getState ) =>
    {
        try 
        {
            const { uid } = getState().auth;

            if( !note.url )
            {
                delete note.url;
            };
    
            const noteToFireStore = { ...note };
            delete noteToFireStore.id; // elimina una propiedad del objeto
    
            await db.doc( `${ uid }/journal/notes/${ note.id }` ).update( noteToFireStore );
    
            dispatch( refreshNote( note.id, noteToFireStore ) );
    
            Swal.fire("Saved",note.title,"success");

        } 
        catch ( error ) 
        {
            Swal.fire("Error",error,"error");
        }
      
    };
};

//////

const refreshNote = ( id, note ) => ( { type : types.notesUpdated, payload : { id, note : { id, ...note } } } );

///////

const startUploadingPicture = ( file ) =>
{
    return async ( dispatch, getState ) =>
    {

        try 
        {
            const { active:activeNote } = getState().notes;

            Swal.fire(
    
                {
                    title: "Uploading....",
                    text: "Please wait...",
                    showConfirmButton: false,
                    onBeforeOpen: ()=>{
                        
                        Swal.showLoading();
                    }
                }
    
            );
    
            const fileUrl = await HFileUpload( file );
            activeNote.url = fileUrl;
            
            dispatch( saveNote( activeNote ) );
    
            Swal.close();
        } 
        catch (error) 
        {
            console.log( error );
        }
       
    };
};

//////

const startDeleteNote = ( id ) =>
{
    return async ( dispatch, getState ) =>
    {
        try 
        {
            const uid = getState().auth.uid;

            await db.doc( `${ uid }/journal/notes/${ id }` ).delete();
    
            dispatch( deleteNote( id ) );
        } 
        catch (error) 
        {
            console.log( error );
        }
       
    };
};

/////

const deleteNote = ( id ) => ( { type : types.notesDelete, payload : id } );

/////

const purgeLogout = () => ( { type : types.notesLogoutCleaning } );


/////

export 
{ 
    starNewNote, 
    activeNote, 
    setNotes, 
    startLoadingNotes, 
    saveNote, 
    refreshNote, 
    startUploadingPicture,
    startDeleteNote,
    purgeLogout
};

