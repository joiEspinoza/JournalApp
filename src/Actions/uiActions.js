import types from "../Types/types";

///////

const setErrorAction = ( error ) => ( {  type :  types.uiSetError, payload : error } );

const uiRemoveErrorAction = () => ( { type : types.uiRemoveError } );

const uiStartLoadingAction = () => ( { type : types.uiStartLoading, payload : true } );

const uiFinishLoading = () => ( { type : types.uiFinishLoading, payload : false } );


///////

export { setErrorAction,uiRemoveErrorAction, uiStartLoadingAction, uiFinishLoading };