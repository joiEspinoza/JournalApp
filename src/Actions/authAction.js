import types from "../Types/types";
import { firebase, googleAuthProvider } from "../FireBase/fireBaseConfig";
import { uiStartLoadingAction, uiFinishLoading } from "./uiActions";
import Swal from 'sweetalert2';
import { purgeLogout } from "./notesAction";

////

const login = ( uid, displayName ) => ( { type : types.login, payload : { uid, displayName } } );

///// LOGIN NORMAL

const starLoginEmailPassword = ( email, password ) =>
{
    return ( dispatch ) => 
    {
        dispatch( uiStartLoadingAction() );

        return firebase.auth().signInWithEmailAndPassword( email, password )
        .then( ( { user } ) => 
        { 
            dispatch( login( user.uid, user.displayName ) );

            dispatch( uiFinishLoading() );

        } )
        .catch( ( error ) => 
        { 
            console.log( error );
            
            dispatch( uiFinishLoading() );

            Swal.fire( "error", error.message, "error" );
            
        } );
    };
};

///// CREAR USUARIO

const starRegisterWhitEmailNamePassword = ( email, password, name ) =>
{
    return( ( dispatch ) => 
    {

        firebase.auth().createUserWithEmailAndPassword( email, password )
        .then( async ( { user } ) =>
        {  
            await user.updateProfile( { displayName : name } ); 
            dispatch( login( user.uid, user.displayName ) );

        } )
        .catch( ( error ) => {
            
            console.log( error );

            Swal.fire( "error", error.message, "error" );
        
        } );

    });
};

///// LOGIN GOOGLE

const startGoogleLogin = () =>
{
    return ( dispatch ) => 
    {
        firebase.auth().signInWithPopup( googleAuthProvider )
        .then( ( { user } ) => dispatch( login( user.uid, user.displayName ) ) )
        .catch( ( error ) => console.log( error ) );
    };
};

/////////////////////////////////////////////----LOGOUT----///////////////////////////////////////

const startLogout = () =>
{
    return async ( dispatch ) => 
    {
        try 
        {
            await firebase.auth().signOut();

            dispatch( purgeLogout() );
            dispatch( logout() );

        } 
        catch( error ) 
        {
            console.log( error );
        };
       
    };
};

const logout = () => ( { type : types.logout } );

//////

export 
{ 
    login, 
    starLoginEmailPassword, 
    startGoogleLogin, 
    starRegisterWhitEmailNamePassword,
    startLogout,
    logout 
};