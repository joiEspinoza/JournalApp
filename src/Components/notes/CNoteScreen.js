import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote, startDeleteNote } from '../../Actions/notesAction';
import useForm from '../../Hooks/useForm';
import NotesAppBar from './NotesAppBar';


////

const CNoteScreen = () => 
{

    const dispatch = useDispatch();

    /////

    const { active:noteActive } = useSelector( state => state.notes );

    const [ formValues, handleInputChange, reset ] = useForm( noteActive );

    const { title, body, id } = formValues;

    /////

    const activeId = useRef( noteActive.id );

    useEffect( () =>  //ayuda a resetear los valores de la nota segun el id | se dispara solo si el ID cambia
    {
        if( noteActive.id !== activeId.current)
        {
            reset( noteActive );
            activeId.current = noteActive.id;
        };

    }, [ noteActive, reset ] );

    ///// 

    useEffect( () => 
    {
       
        dispatch( activeNote( formValues.id, { ...formValues } ) );

    }, [ formValues, dispatch ] );

    //////
    
    const handleDelete = () =>
    {
        dispatch( startDeleteNote( id ) );
    };

    /////


    return (

        <div className="notes__main-content">

            <NotesAppBar/>

            <div className="notes__content">

                <input type="text" placeholder="Some awesone title" autoComplete="off" className="notes__title-input" name="title" value={ title } onChange={ handleInputChange }/>

                <textarea placeholder="What happen today?" className="notes__textarea" name="body" value={ body } onChange={ handleInputChange } ></textarea>
                
                { (noteActive.url) && 
                
                    (<div className="notes__image">

                            <img 
                                src={ noteActive.url }
                                alt="img"
                            />

                    </div>)
                }

            </div>

            <button className="btn btn-danger" onClick={ handleDelete }> Delete </button>
            
        </div>

    );

};

/////

export default CNoteScreen;
