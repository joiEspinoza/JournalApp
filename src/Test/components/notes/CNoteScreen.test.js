const { mount } = require("enzyme");
const { Provider } = require("react-redux");
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from "redux-thunk";
import { activeNote } from '../../../Actions/notesAction';
import CNoteScreen from '../../../Components/notes/CNoteScreen';

///////

jest.mock( '../../../Actions/notesAction', () => ( { 
    
    activeNote : jest.fn(), 

} ) );

///////

const middlewares = [ thunk ]
const mockStore = configureStore( middlewares );

////////

const initState = 
{ 
    auth : { uid : "ABCD1234", name : "TestMan" }, 
    ui: { loading : false, msgError : null }, 
    notes : { note : [], active : { id:"ABC123", title : "", body : "esto es body", date : 0 } } 
};

let store = mockStore( initState );
store.dispatch = jest.fn();

////////

    const wrapper = mount( 
    
        <Provider store={ store } >
            <CNoteScreen/> 
        </Provider>
    
    );

    ///////

describe('Pruebas en CNoteScreen.js', () => 
{
   test('Debe mostrarse correctamente ', () => 
   {
       
        expect( wrapper ).toMatchSnapshot();


   });

   //////

   test('Debe disparar el activeNote ', () => 
   {
      
        wrapper.find( 'input[name="title"]' ).simulate( "change", { target : { value : "Test Title", name : "title" } } );

      
        expect( activeNote ).toHaveBeenCalledWith( "ABC123", { id : "ABC123", title : "Test Title", body : "esto es body", date : 0 } );

   });
   
    
});
