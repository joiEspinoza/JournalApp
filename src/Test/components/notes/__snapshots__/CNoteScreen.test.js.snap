// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en CNoteScreen.js Debe mostrarse correctamente  1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      25-11-2020
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesone title"
      type="text"
      value=""
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="What happen today?"
      value="esto es body"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
     Delete 
  </button>
</div>
`;
